src/models.py

```diff
@@ -0,0 +1,439 @@
+import psycopg2
+from api import db
+from aux import logger_instance
+from errors import Error, FATAL
+
+logger = logger_instance(__name__)
+
+
+class Event(object):
+    def __init__(self, event_id, work_uri, measure_uri, timestamp, value,
+                 event_uri=None, country_uri=None, uploader_uri=''):
+        self.event_id  = event_id
+        self.work_uri = work_uri
+        self.measure_uri = measure_uri
+        self.timestamp = timestamp.strftime('%Y-%m-%dT%H:%M:%S%z')
+        self.value = int(value)
+        self.event_uri = event_uri
+        self.country_uri = country_uri
+        self.uploader_uri = uploader_uri
+
+    def save(self):
+        try:
+            q = '''INSERT INTO event (event_id, work_uri, measure_uri,
```

Comment:
ORM's are fine for simple queries (e.g. `select * from x;`), and people tend to like them because they're a 'safe' wrapper, but the fact that they're a wrapper means they don't have all the functionality SQL has and they can also have unexpected side effects - I know 'raw SQL' is not very popular, but it's a lot safer to execute an SQL query you have written than having some random library in language x attempt to translate to SQL.

Anyway, in this particular example no magic wrapper is used as they don't support the `ON CONFLICT DO NOTHING` statement

Reply:
I will have to contradict you on 3 different points:

* `they don't have all the functionality SQL has` this is true for contexts that are *way* more complex than the one you have here, and for very small niches that we will never touch in this projects
* `I know 'raw SQL' is not very popular, but it's a lot safer to execute an SQL query` false; this exposes you to different kinds of security issues (SQL injections to mention one)
* SQLAlchemy supports `ON CONFLICT`: https://docs.sqlalchemy.org/en/13/dialects/postgresql.html#insert-on-conflict-upsert

Reply:
You don't need an ORM to avoid security issues (neither do you need it to construct your queries) - connection, pointer handling, and input parsing is done at the database API layer (Psycopg in this case).

It's good to know that sqlalchemy does support on conflict statements, but honestly I don't see any advantage whatsoever in using an ORM to construct SQL queries; we already speak SQL! SQL queries are a lot easier to read than concatenated methods that are specific to one particular library.

Reply:
Indeed, Psychopg will protect you from SQL injection.

Not using an ORM is at the same time:

* *increasing the overhead for future maintenance*: we will end up with millions of rows in this DB, and we will have no easy way to create DB migrations if we want to change the table structure (i.e. no Alemic, no South); this is fine if this project is going to be delivered and forgotten, but since we will run this in our production environment, and we are aiming for long term maintenance, not using an ORM now will increase maintenance cost in the future.

* *limiting flexibility on handling of DB connection*: a lot of Psycopg features are not available / configurable using the web.py adapter (`web.database`); i.e. we don't have any option to configure connection pooling (which is available in Psycopg but not in web.py -- and it is available in most ORMs)

* most open source projects _not_ using an ORM do so because they need to define their own transaction logic ([Odoo is a good example](https://github.com/odoo/odoo/blob/8f4b98e67fe49516c0c830524de351db67c58357/odoo/sql_db.py#L68), where the transaction logic is so critical that they couldn't rely on an ORM); we are not aiming for this level of complexity in this project, and we are self-imposing limits on future maintainability for no real reason. A quick search on the most popular open source projects on Github will show that 96% of them are using an ORM or another, since it adds flexibility without sacrificing performance.

* [Sentry](https://github.com/getsentry/sentry/tree/master/src/sentry/db) is based on the Django ORM
* [pyspider](https://github.com/binux/pyspider/blob/3fb9167983f0123b9cac6615e543e9fbaf68cc04/pyspider/database/sqlalchemy/projectdb.py) SQLAlchemy

Reply:
We need to differentiate the database (and its schema) from this repository completely. There is no ORM here, and definitely no migrations as we're not going to let the code dictate the database schema. Any changes to the schema will occur in https://github.com/hirmeos/metrics_db and will need to be taken into serious consideration before applying them as they will require an update not just to this repo but everywhere else - hence our insistence in getting the data model right from the very beginning. The fact that we will have millions of rows using a normalised data model does not mean we should impose writing SQL with a particular library - if we get to a point in which any changes to the schema need to take place we will write the necessary SQL to do so.

I'm up for replacing webpy's wrapper with the native Psycopg if there's a benefit to it (it will take very little time to replace the ~15 statements). But pooling in this case is not needed - there's a single database connection that doesn't get closed.inin

I honestly was expecting a review of the schema or the queries available, and I had simply added Rowan as a reviewer to check I hadn't broken any of the private CI stuff that has been hardcoded in or the random listener that has been plugged in and will spit out a log line every 30 seconds if you don't run the altmetrics service yourself - these are the sort of things we *do* need to fix as they impose unneeded dependencies on anyone wanting to run a standalone API that only depends on a self-defined datbase. We have, however, decided to postpone that review to a more pertinent time as what we need now is a running API that respects the schema.

I'm on leave today and tomorrow, but we already have this branch running in production with all our own and OAPEN's data - so if you can't or don't want to deploy this we can point people to our version and start populating it for validation of the workpackage.

Reply:
I am happy to close this conversation specifically for the sake of delivering the work package, but after the delivery phase, we will face a maintenance phase for the next 12 months (at least), and this will need to be accounted for when planning the maintenance costs for the service. In 2019 writing raw SQL for DB migrations is not advisable, especially for such a simple application.

Rowan will give his own review.

In the past we discussed the absolute necessity to add tests to our services, that will need to be in place before we can run this in production.